libraries/AP_Scripting/AP_Scripting_helpers.cpp

```diff
@@ -195,4 +195,316 @@ bool Parameter::set_default(float value)
+        return true;
+    }
+
+    payload.length = transfer.payload_len;
+    payload.node_id = transfer.source_node_id;
+    payload.timestamp = transfer.timestamp_usec;
+#if CANARD_ENABLE_CANFD
+    payload.canfd = transfer.canfd;
+#endif
+
+    uint32_t bits_remaining = transfer.payload_len * 8;
+    uint32_t ofs = 0;
+
+    // decode in bytes, we don't have a canardDecodeBuffer function
+    while (bits_remaining > 0) {
+        uint8_t nbits = MIN(8U, bits_remaining);
+        canardDecodeScalar(&transfer, ofs*8, nbits, false, (void*)&payload.data[ofs]);
+        ofs++;
+        bits_remaining -= nbits;
+    }
+
+    // push to queue
+    payloads.push(payload);
+
+    return true;
+}
+
+/*
+  check for an incoming message
+ */
+int DroneCAN_Handle::check_message(lua_State *L)
+{
+    /*
+      get the receive mutex for this driver
+     */
+    DroneCAN_Handle *h = check_DroneCAN_Handle(L, 1);
+    WITH_SEMAPHORE(h->dc->get_canard_iface().get_sem_rx());
+
+    auto *s = h->subscriber;
+
+    if (s == nullptr) {
+        return 0;
+    }
+    Subscriber::Payload payload;
+    if (!s->payloads.pop(payload)) {
+        return 0;
+    }
+
+    lua_pushlstring(L, (char *)payload.data, payload.length);
+    free(payload.data);
+
+    lua_pushinteger(L, payload.node_id);
+    *new_uint64_t(L) = payload.timestamp;
+    lua_pushboolean(L, payload.canfd);
+
+    if (s->trans_type == CanardTransferTypeResponse) {
+        // request reply removes the subscriber
+        delete h->subscriber;
+        h->subscriber = nullptr;
+    }
+    
+    return 4;
+}
+
+/*
+  cleanup any dangling subscribers
+ */
+void DroneCAN_Handle::destroy_all(void)
+{
+    /*
+      we need to stop any processing from happening while we cleanup
+     */
+    for (uint8_t i=0; i<CANARD_NUM_HANDLERS; i++) {
+        auto *dc = AP_DroneCAN::get_dronecan(i);
+        if (dc != nullptr) {
+            dc->get_canard_iface().get_sem_rx().take_blocking();
+        }
+    }
+    while (Subscriber::cleanup) {
+        delete Subscriber::cleanup;
+    }
+    for (uint8_t i=0; i<CANARD_NUM_HANDLERS; i++) {
+        auto *dc = AP_DroneCAN::get_dronecan(i);
+        if (dc != nullptr) {
+            dc->get_canard_iface().get_sem_rx().give();
+        }
+    }
+}
+
+/*
+  garbage collect a handle
+ */
+int DroneCAN_Handle::__gc(lua_State *L)
+{
+    DroneCAN_Handle *h = check_DroneCAN_Handle(L, 1);
+    WITH_SEMAPHORE(h->dc->get_canard_iface().get_sem_rx());
+
+    if (h->subscriber != nullptr) {
+        delete h->subscriber;
+        h->subscriber = nullptr;
+    }
+    return 0;
+}
+
+// lua constructor for DroneCAN_Handle
+int DroneCAN_Handle::new_handle(lua_State *L)
+{
+    static bool metatable_created;
+    if (!metatable_created) {
+        metatable_created = true;
+        luaL_newmetatable(L, "DroneCAN_Handle");
+    
+        lua_pushcfunction(L, DroneCAN_Handle::__gc);
+        lua_setfield(L, -2, "__gc");
+    
+        lua_pop(L, 1);
+    }
```

Comment:
```suggestion

```

This can be handled by the binding generator with `userdata DroneCAN_Handle manual_operator __gc DroneCAN_Handle::__gc`.

Reply:
i thought I'd tried this, but I just tried it again and it does work, thanks! changed